File: NET/Libraries/Treehopper.Libraries/Sensors/Inertial/Mpu9250.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Numerics;
using System.Threading.Tasks;
using Treehopper.Libraries.Sensors.Magnetic;

namespace Treehopper.Libraries.Sensors.Inertial
{
    /// <summary>
    ///     InvenSense MPU9250 9-DoF IMU
    /// </summary>
    [Supports("InvenSense", "MPU-9250")]
    public class Mpu9250 : Mpu6050, IMagnetometer
    {
        /// <summary>
        ///     The 3-axis magnetometer data
        /// </summary>
        protected Vector3 magnetometer;
        Ak8975 mag;

        public static IList<Mpu9250> Probe(I2C i2c)
        {
            var retVal = new List<Mpu9250>();
            try
            {
                var dev = new SMBusDevice(0x68, i2c, 100);
                var whoAmI = Task.Run(() => dev.ReadByteData(0x75)).Result;
                if (whoAmI == 0x71)
                    retVal.Add(new Mpu9250(i2c, false));
            } catch(Exception ex) { }

            try
            {
                var dev = new SMBusDevice(0x69, i2c, 100);
                var whoAmI = Task.Run(() => dev.ReadByteData(0x75)).Result;
                if (whoAmI == 0x71)
                    retVal.Add(new Mpu9250(i2c, true));
            }
            catch (Exception ex) { }

            return retVal;
        }

        public Mpu9250(I2C i2c, bool addressPin = false, int rate = 400) : base(i2c)
        {
            mag = new Ak8975(i2c);
            mag.AutoUpdateWhenPropertyRead = false;
            _registers.intPinCfg.bypassEn = 1;
            _registers.intPinCfg.latchIntEn = 1;
            Task.Run(_registers.intPinCfg.write).Wait();
        }

        /// <summary>
        ///     Whether the magnetometer is enabled or not
        /// </summary>
        public bool EnableMagnetometer { get; set; } = true;

        /// <summary>
        ///     The current 3-axis magnetometer data
        /// </summary>
        public Vector3 Magnetometer
        {
            get
            {
                if (AutoUpdateWhenPropertyRead) Update().Wait();
                return magnetometer;
            }
        }

        /// <summary>
        ///     Retrieve the latest sample data from the MPU9250
        /// </summary>
        /// <returns></returns>
        public override async Task Update()
        {
            await base.Update().ConfigureAwait(false);

            if (!EnableMagnetometer) return;
            await mag.Update().ConfigureAwait(false);
            magnetometer = mag.Magnetometer;
        }
    }
}

File: Java/api/treehopper.libraries/src/main/java/io/treehopper/libraries/sensors/inertial/mpu6050/Mpu9250.java
package io.treehopper.libraries.sensors.inertial.mpu6050;

import com.badlogic.gdx.math.Vector3;
import io.treehopper.interfaces.I2c;
import io.treehopper.libraries.sensors.magnetic.IMagnetometer;
import io.treehopper.libraries.sensors.magnetic.ak8975.Ak8975;

public class Mpu9250 extends Mpu6050 implements IMagnetometer {
    private Ak8975 mag;
    private boolean enableMagnetometer = true;

    public Mpu9250(I2c i2c, boolean addressPin, int rate) {
        super(i2c, addressPin, rate);
        mag = new Ak8975(i2c);
        mag.setAutoUpdateWhenPropertyRead(false);
        registers.intPinCfg.bypassEn = 1;
        registers.intPinCfg.latchIntEn = 1;
        registers.intPinCfg.write();
    }

    @Override
    public void update() {
        super.update();

        if (!enableMagnetometer) return;
        mag.update();
    }

    public boolean isEnableMagnetometer() {
        return enableMagnetometer;
    }

    public void setEnableMagnetometer(boolean enableMagnetometer) {
        this.enableMagnetometer = enableMagnetometer;
    }

    @Override
    public Vector3 getMagnetometer() {
        if (isAutoUpdateWhenPropertyRead())
            update();

        return mag.getMagnetometer();
    }
}
